src/app/clusters/thermostat-server/thermostat-server.cpp

```diff
@@ -220,6 +220,293 @@ int16_t EnforceCoolingSetpointLimits(int16_t CoolingSetpoint, EndpointId endpoin
+{
+    if (!autoSupported)
+    {
+        return Status::Success;
+    }
+    int16_t minValidCoolingSetpoint = newHeatingSetpoint + deadband;
+    if (minValidCoolingSetpoint > maxCoolingSetpoint)
+    {
+        // If we need to adjust the cooling setpoint to preserve the deadband, it will go above the max cool setpoint
+        return Status::InvalidValue;
+    }
+    // It's possible to adjust the cooling setpoint, if needed
+    return Status::Success;
+}
+
+typedef Status (*setpointSetter)(EndpointId endpoint, int16_t value, MarkAttributeDirty markDirty);
+
+/**
+ * @brief Attempts to ensure that a change to the heating setpoint maintains the deadband with the cooling setpoint
+ * by adjusting the cooling setpoint
+ *
+ * @param endpoint The endpoint on which the heating setpoint has been changed
+ * @param currentCoolingSetpoint The current cooling setpoint
+ * @param newHeatingSetpoint The newly adjusted heating setpoint
+ * @param maxCoolingSetpoint The maximum allowed cooling setpoint
+ * @param deadband The deadband to preserve
+ * @param setter A function for setting the cooling setpoint
+ */
+void ensureCoolingSetpointDeadband(EndpointId endpoint, int16_t currentCoolingSetpoint, int16_t newHeatingSetpoint,
+                                   int16_t maxCoolingSetpoint, int16_t deadband, setpointSetter setter)
+{
+    int16_t minValidCoolingSetpoint = newHeatingSetpoint + deadband;
+    if (currentCoolingSetpoint >= minValidCoolingSetpoint)
+    {
+        // The current cooling setpoint doesn't violate the deadband
+        return;
+    }
+    if (minValidCoolingSetpoint > maxCoolingSetpoint)
+    {
+        // Adjusting the cool setpoint to preserve the deadband would violate the max cool setpoint
+        // This should have been caught in checkCoolingSetpointDeadband, so log and exit
+        ChipLogError(Zcl, "Failed ensuring cooling setpoint deadband");
+        return;
+    }
+    // Adjust the cool setpoint to preserve deadband
+    auto status = setter(endpoint, minValidCoolingSetpoint, MarkAttributeDirty::kYes);
+    if (status != Status::Success)
+    {
+        ChipLogError(Zcl, "Error: ensureCoolingSetpointDeadband failed!");
+    }
+}
+
+/**
+ * @brief Attempts to ensure that a change to the cooling setpoint maintains the deadband with the heating setpoint
+ * by adjusting the heating setpoint
+ *
+ * @param endpoint The endpoint on which the cooling setpoint has been changed
+ * @param currentHeatingSetpointThe current heating setpoint
+ * @param newCoolingSetpoint The newly adjusted cooling setpoint
+ * @param minHeatingSetpoint The minimum allowed cooling setpoint
+ * @param deadband The deadband to preserve
+ * @param setter A function for setting the heating setpoint
+ */
+void ensureHeatingSetpointDeadband(EndpointId endpoint, int16_t currentHeatingSetpoint, int16_t newCoolingSetpoint,
+                                   int16_t minHeatingSetpoint, int16_t deadband, setpointSetter setter)
```

Comment:
`SetpointSetter` type should have capital `S` at the start

Reply:
Fixed.